src/zenml/code_repositories/git/local_git_repository.py

```diff
@@ -0,0 +1,169 @@
+    from git.objects import Commit
+    from git.remote import Remote
+    from git.repo.base import Repo
+
+logger = get_logger(__name__)
+
+
+class LocalGitRepository(LocalRepository):
+    """Local git repository."""
+
+    def __init__(
+        self, code_repository_id: UUID, git_repo: "Repo", remote_name: str
+    ):
+        """Initializes a local git repository.
+
+        Args:
+            code_repository_id: The ID of the code repository.
+            git_repo: The git repo.
+            remote_name: Name of the remote.
+        """
+        super().__init__(code_repository_id=code_repository_id)
+        self._git_repo = git_repo
+        self._remote = git_repo.remote(name=remote_name)
+
+    @classmethod
+    def at(
+        cls,
+        path: str,
+        code_repository_id: UUID,
+        remote_url_validation_callback: Callable[[str], bool],
+    ) -> Optional["LocalGitRepository"]:
+        """Returns a local git repository at the given path.
+
+        Args:
+            path: The path to the local git repository.
+            code_repository_id: The ID of the code repository.
+            remote_url_validation_callback: A callback that validates the
+                remote URL of the git repository.
+
+        Returns:
+            A local git repository if the path is a valid git repository
+            and the remote URL is valid, otherwise None.
+        """
+        try:
+            # These imports fail when git is not installed on the machine
+            from git.exc import InvalidGitRepositoryError
+            from git.repo.base import Repo
+        except ImportError:
+            return None
+
+        try:
+            git_repo = Repo(path=path, search_parent_directories=True)
+        except InvalidGitRepositoryError:
+            return None
+
+        remote_name = None
+        for remote in git_repo.remotes:
+            if remote_url_validation_callback(remote.url):
+                remote_name = remote.name
+                break
+
+        if not remote_name:
+            return None
+
+        return LocalGitRepository(
+            code_repository_id=code_repository_id,
+            git_repo=git_repo,
+            remote_name=remote_name,
+        )
+
+    @property
+    def git_repo(self) -> "Repo":
+        """The git repo.
+
+        Returns:
+            The git repo object of the local git repository.
+        """
+        return self._git_repo
+
+    @property
+    def remote(self) -> "Remote":
+        """The git remote.
+
+        Returns:
+            The remote of the git repo object of the local git repository.
+        """
+        return self._remote
+
+    @property
+    def root(self) -> str:
+        """The root of the git repo.
+
+        Returns:
+            The root of the git repo.
+        """
+        assert self.git_repo.working_dir
+        return str(self.git_repo.working_dir)
+
+    @property
+    def is_dirty(self) -> bool:
+        """Whether the git repo is dirty.
```

Comment:
Wondering if it's worth defining 'dirty' here? (Unless this is a commonly used term, in which case feel free to ignore, but it seemed like this is a ZenML-specific use?)

Reply:
It's a commonly used git term, but I'll add some clarification anyway